src/data_structures/suffix_array.rs

```diff
@@ -19,12 +20,79 @@ use vec_map::VecMap;
 
 use alphabets::{Alphabet, RankTransform};
 use data_structures::smallints::SmallInts;
+use data_structures::bwt::{BWT, Less, Occ};
 
+pub trait SuffixArray {
+    fn get(&self, index: usize) -> Option<usize>;
+    fn range(&self, range: Range<usize>) -> Option<Vec<usize>>;
```

Comment:
I played around with it some, and I couldn't get it (or anything similar) working quite right. Your suggestion doesn't work for a couple of reasons. The first is that [usize] needs to be bounded by the lifetime of the suffix array, while Vec<usize> cannot be. The second is that [usize] doesn't implement Deref for [usize] ('doh-- really, shouldn't everything be a Deref for itself? Anyway...).

I got a workaround using an enum. I had to do some silly things like implementing my own PartialEq, Eq, and conversion to iterator, but otherwise, it seems like it's working okay. I was surprised to find there's not a common List/Array-like trait implemented by both. It seems like the recently added RFC for specialization would make something like that more feasible/performant. Again though, I'm getting off topic.